index.d.ts

```diff
@@ -1,288 +1,615 @@
+  export namespace useGroupBy {
+    export const pluginName = 'useGroupBy'
   }
 
-  export interface UseGroupByValues {
-    groupByFn: any
+  export type UseGroupByOptions<D extends object> = Partial<{
     manualGroupBy: boolean
     disableGrouping: boolean
-    aggregations: any
+    aggregations: Record<string, AggregatorFn<D>>
+    groupByFn: (rows: Row<D>[], columnId: IdType<D>) => Record<string, Row<D>>
+  }>
+
+  export interface UseGroupByHooks<D extends object> {
+    getGroupByToggleProps: ((
+      header: HeaderGroup<D>,
+      instance: TableInstance<D>
+    ) => object)[]
+  }
+
+  export interface UseGroupByState<D extends object> {
+    groupBy: IdType<D>[]
+  }
+
+  export type UseGroupByColumnOptions<D extends object> = Partial<{
+    aggregate: Aggregator<D> | Aggregator<D>[]
+    Aggregated: Renderer<CellProps<D>>
+    disableGrouping: boolean
+    groupByBoundary: boolean
+  }>
+
+  export interface UseGroupByInstanceProps<D extends object> {
+    rows: Row<D>[]
+    preGroupedRows: Row<D>[]
+    toggleGroupBy: (columnId: IdType<D>, toggle: boolean) => void
   }
 
-  export interface UseGroupByState {
-    groupBy: string[]
-    isAggregated?: boolean
+  export interface UseGroupByColumnProps<D extends object> {
+    canGroupBy: boolean
+    isGrouped: boolean
+    groupedIndex: number
+    toggleGroupBy: () => void
   }
 
-  export interface UseExpandedValues {
-    toggleExpanded?: () => any
+  export interface UseGroupByHeaderProps<D extends object> {
+    getGroupByToggleProps: (props?: object) => object
   }
 
-  export interface UseSortbyOptions {
-    sortByFn?: SortByFn
-    manualSorting?: boolean
-    disableSorting?: boolean
-    defaultSortDesc?: boolean
-    disableMultiSort?: boolean
+  export interface UseGroupByRowProps<D extends object> {
+    isAggregated: boolean
+    groupByID: IdType<D>
+    groupByVal: string
+    values: Record<IdType<D>, AggregatedValue>
+    subRows: Row<D>[]
+    depth: number
+    path: IdType<D>[]
+    index: number
   }
 
-  export interface UseSortbyState<D> {
-    sortBy?: SortingRules<D>
+  export interface UseGroupByCellProps<D extends object> {
+    isGrouped: boolean
+    isRepeatedValued: boolean
+    isAggregated: boolean
   }
 
-  export interface TableInstance<D = {}> extends TableOptions<D> {
-    hooks: Hooks
+  export type DefaultAggregators =
+    | 'sum'
+    | 'average'
+    | 'median'
+    | 'uniqueCount'
+    | 'count'
+
+  export type AggregatorFn<D extends object> = (
+    columnValues: CellValue[],
     rows: Row<D>[]
-    columns: EnhancedColumn<D>[]
-    headerGroups: HeaderGroup<D>[]
-    headers: HeaderGroup<D>[]
-    getTableProps: (userProps?: any) => any
-    getRowProps: (userProps?: any) => any
-    prepareRow: (row: Row<D>) => any
+  ) => AggregatedValue
+  type Aggregator<D extends object> =
+    | AggregatorFn<D>
+    | DefaultAggregators
+    | string // TODO: Check that DefaultAggregators works
+  type AggregatedValue = unknown
```

Comment:
I'm unsure about using unknown in this and a couple of other places, on one hand it is technically correct, but it sure causes some ugly casts all over the places.  Making it be any instead makes using this much easier.

Reply:
Yeah, I did notice that happening which is why some things I have switched over so far. (ie. `CellValue` and `FilterValue`) I will keep playing with the example repo and see if it becomes and issue.